File: stock-market-service/src/main/java/rs/edu/raf/service/impl/OrderServiceImpl.java
package rs.edu.raf.service.impl;

import lombok.AllArgsConstructor;
import org.springframework.http.HttpMethod;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import rs.edu.raf.dto.*;
import rs.edu.raf.exceptions.PlaceOrderException;
import rs.edu.raf.mapper.OrderMapper;
import rs.edu.raf.model.ListingOwner;
import rs.edu.raf.model.Stock;
import rs.edu.raf.model.order.Order;
import rs.edu.raf.model.order.OrderAction;
import rs.edu.raf.repository.ListingOwnerRepository;
import rs.edu.raf.repository.OrderRepository;
import rs.edu.raf.repository.StockRepository;
import rs.edu.raf.service.OrderService;
import rs.edu.raf.strategy.OrderProcessStrategy;
import rs.edu.raf.strategy.OrderProcessStrategyFactory;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class OrderServiceImpl implements OrderService {
    private OrderRepository orderRepository;
    private OrderMapper orderMapper;
    private OrderProcessStrategyFactory orderProcessStrategyFactory;
    private ListingOwnerRepository listingOwnerRepository;
    private RestTemplate bankServiceRestTemplate;
    private StockRepository stockRepository;

    @Override
    public List<OrderDTO> getAllOrders() {
        return orderRepository.findAll().stream().map(orderMapper::orderToOrderDTO).toList();
    }

    @Override
    public OrderDTO createMarketOrder(MarketOrderCreateDTO marketOrderCreateDTO) {
        return placeOrder(orderMapper.marketOrderCreateDTOtoOrder(marketOrderCreateDTO));
    }

    @Override
    public OrderDTO createLimitOrder(LimitOrderCreateDTO limitOrderCreateDTO) {
        return placeOrder(orderMapper.limitOrderCreateDTOtoOrder(limitOrderCreateDTO));
    }

    @Override
    public OrderDTO createStopOrder(StopOrderCreateDTO stopOrderCreateDTO) {
        return placeOrder(orderMapper.stopOrderCreateDTOtoOrder(stopOrderCreateDTO));
    }

    @Override
    public OrderDTO createStopLimitOrder(StopLimitOrderCreateDTO stopLimitOrderCreateDTO) {
        return placeOrder(orderMapper.stopLimitOrderCreateDTOtoOrder(stopLimitOrderCreateDTO));
    }

    @Override
    public OrderDTO createAllOrNoneOrder(AllOrNoneOrderCreateDTO allOrNoneOrderCreateDTO) {
        return placeOrder(orderMapper.allOrNoneOrderCreateDTOtoOrder(allOrNoneOrderCreateDTO));
    }

    @Override
    public boolean deductBalanceForOrder(Order order, Double amount) {
        return bankServiceRestTemplate.exchange("/bank-accounts/deduct-available-balance/" + order.getOrderCreator()
                        + "/" + amount, HttpMethod.PUT, null , Boolean.class).getBody();
    }

    @Override
    public void processOrders() {
        for (Order order: orderRepository.findAllByQuantityGreaterThan(0)) {
            OrderProcessStrategy orderProcessStrategy = orderProcessStrategyFactory.getExecutionStrategy(order);
            orderProcessStrategy.processOrder(order);
        }
    }

    @Override
    public OrderDTO placeOrder(Order order) {
        Stock stock = stockRepository.findById(order.getTicker()).orElseThrow();

        boolean canPlaceBuyOrder = order.getOrderAction() == OrderAction.BUY
                && deductBalanceForOrder(order, order.getQuantity() * stock.getPrice());

        boolean canPlaceSellOrder = order.getOrderAction() == OrderAction.SELL && canPlaceSellOrder(order);

        if(canPlaceBuyOrder || canPlaceSellOrder) {
            return orderMapper.orderToOrderDTO(orderRepository.save(order));
        }

        throw new PlaceOrderException();
    }

    @Override
    public boolean canPlaceSellOrder(Order order) {
        Optional<ListingOwner> optionalListingOwner =
                listingOwnerRepository.findByOwnerAndTicker(order.getOrderCreator(), order.getTicker());

        if(optionalListingOwner.isEmpty()) {
            return false;
        }

        return orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(order.getTicker(), order.getOrderAction(), 0)
                .stream()
                .map(Order::getQuantity)
                .reduce(0, Integer::sum) + order.getQuantity() <= optionalListingOwner.get().getQuantity();
    }
}

File: stock-market-service-kotlin/src/main/kotlin/org/example/stockmarketservicekotlin/service/impl/OrderServiceImpl.kt
package org.example.stockmarketservicekotlin.service.impl

import org.example.stockmarketservicekotlin.dto.*
import org.example.stockmarketservicekotlin.exceptions.PlaceOrderException
import org.example.stockmarketservicekotlin.mapper.OrderMapper
import org.example.stockmarketservicekotlin.model.ListingOwner
import org.example.stockmarketservicekotlin.model.Stock
import org.example.stockmarketservicekotlin.model.order.Order
import org.example.stockmarketservicekotlin.model.order.OrderAction
import org.example.stockmarketservicekotlin.repository.ListingOwnerRepository
import org.example.stockmarketservicekotlin.repository.OrderRepository
import org.example.stockmarketservicekotlin.repository.StockRepository
import org.example.stockmarketservicekotlin.service.OrderService
import org.example.stockmarketservicekotlin.strategy.OrderProcessStrategy
import org.example.stockmarketservicekotlin.strategy.OrderProcessStrategyFactory
import org.springframework.http.HttpMethod
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.util.*

@Service
class OrderServiceImpl(
    private val orderRepository: OrderRepository,
    private val orderMapper: OrderMapper,
    private val orderProcessStrategyFactory: OrderProcessStrategyFactory,
    private val listingOwnerRepository: ListingOwnerRepository,
    private val bankServiceRestTemplate: RestTemplate,
    private val stockRepository: StockRepository,
) : OrderService {
    override fun getAllOrders(): List<OrderDTO> {
        return orderRepository.findAll().stream().map(orderMapper::orderToOrderDTO).toList()
    }

    override fun createMarketOrder(marketOrderCreateDTO: MarketOrderCreateDTO): OrderDTO {
        return placeOrder(orderMapper.marketOrderCreateDTOtoOrder(marketOrderCreateDTO))
    }

    override fun createLimitOrder(limitOrderCreateDTO: LimitOrderCreateDTO): OrderDTO {
        return placeOrder(orderMapper.limitOrderCreateDTOtoOrder(limitOrderCreateDTO))
    }

    override fun createStopOrder(stopOrderCreateDTO: StopOrderCreateDTO): OrderDTO {
        return placeOrder(orderMapper.stopOrderCreateDTOtoOrder(stopOrderCreateDTO))
    }

    override fun createStopLimitOrder(stopLimitOrderCreateDTO: StopLimitOrderCreateDTO): OrderDTO {
        return placeOrder(orderMapper.stopLimitOrderCreateDTOtoOrder(stopLimitOrderCreateDTO))
    }

    override fun createAllOrNoneOrder(allOrNoneOrderCreateDTO: AllOrNoneOrderCreateDTO): OrderDTO {
        return placeOrder(orderMapper.allOrNoneOrderCreateDTOtoOrder(allOrNoneOrderCreateDTO))
    }

    override fun deductBalanceForOrder(order: Order, amount: Double): Boolean {
        return bankServiceRestTemplate.exchange(
            ("/bank-accounts/deduct-available-balance/" + order.orderCreator + "/" + amount),
            HttpMethod.PUT,
            null,
            Boolean::class.java
        ).body ?: false
    }

    override fun processOrders() {
        for (order in orderRepository.findAllByQuantityGreaterThan(0)) {
            val orderProcessStrategy: OrderProcessStrategy = orderProcessStrategyFactory.getExecutionStrategy(order)
            orderProcessStrategy.processOrder(order)
        }
    }

    override fun placeOrder(order: Order): OrderDTO {
        val stock: Stock = stockRepository.findById(order.ticker).orElseThrow()

        val canPlaceBuyOrder =
            order.orderAction == OrderAction.BUY && deductBalanceForOrder(order, order.quantity * stock.price)

        val canPlaceSellOrder = order.orderAction === OrderAction.SELL && canPlaceSellOrder(order)

        if (canPlaceBuyOrder || canPlaceSellOrder) {
            return orderMapper.orderToOrderDTO(orderRepository.save(order))
        }

        throw PlaceOrderException(order.id)
    }

    override fun canPlaceSellOrder(order: Order): Boolean {
        val optionalListingOwner: Optional<ListingOwner> =
            listingOwnerRepository.findByOwnerAndTicker(order.orderCreator, order.ticker)

        if (optionalListingOwner.isEmpty) {
            return false
        }

        return orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
            order.ticker, order.orderAction, 0
        ).stream().map(Order::quantity)
            .reduce(0) { a: Int, b: Int -> Integer.sum(a, b) } + order.quantity <= optionalListingOwner.get().quantity
    }
}